### ISSUE TITLE
Type check in impossible clause is limited

### ISSUE THREAD

**Issue by locriacyber** [idris-lang/Idris2#2431]
# Steps to Reproduce
Type check the following file.
```idris2
true_false : (True = False) -> Void
true_false (True = 3) impossible
true_false (True = True) impossible
true_false Refl impossible
-- true_false 1 impossible -- compiler unhappy about this one
```
# Expected Behavior
All clauses should not type check, probably.

# Observed Behavior
Idris2 says `true_false` is total.

**Comment by buzden** [comment]
You say that you'd expect clause with `Refl` to not to typecheck. Why? It seems to be perfectly typecheckable.

And what's wrong with observed behaviour that you mention? Such function can be implemented total since it finishes and returns correctly typed result for every possible argument; like, all balls of an empty set are blue.

**Comment by gallais** [comment]
I suppose we could argue that whatever constructor pattern we have at least
belongs to the type family being matched on rather than any old garbage.

The current machinery is not really set up to distinguish an error at the head
of the term vs. a deeper error though.

**Comment by locriacyber** [comment]
> You say that you'd expect clause with `Refl` to not to typecheck. Why? It seems to be perfectly typecheckable.
> 
> And what's wrong with observed behaviour that you mention? Such function can be implemented total since it finishes and returns correctly typed result for every possible argument; like, all balls of an empty set are blue.

`Refl` cannot be the type `(True = False)`. It doesn't make any sense in this context.

**Comment by buzden** [comment]
`impossible` exactly says that those matches before this keyword must not make sense, i.e. cannot be matched. That's why `true_false Refl impossible` typechecks: compiler checks that `Refl` cannot match a value of type `True = False`. At the same time coverage checker checks that you've covered everything that can ever be covered, since nothing except `Refl` may make sense for type `whatever = whatever_else`.

By the way, at least one `impossible` clause allow compiler to distinguish between unimplemented function and implemented uncallable function.

Does it make sense to you now?

**Comment by locriacyber** [comment]
Thank you!

**Comment by buzden** [comment]
> `bb` is total in Idris 2, version 0.5.1-67951529e. Is it still unimplemented function?

If this is all the code that relates to `bb`, then yes, it is unimplemented. Compiler can tell you itself; e.g. in REPL, you can call `:metavars` to see all the holes.

By the way, github issues is not the best place for long conversations and questions. For this, we have a [discord](https://discord.gg/UX68fDs2jc).

**Comment by locriacyber** [comment]
Thank you again! Do I need to close this issue?

**Comment by gallais** [comment]
> Do I need to close this issue?

No, I think it's an interesting issue and we should definitely be stricter in what we accept in
`impossible` clauses.
